src/cfnlint/conditions.py

```diff
@@ -0,0 +1,316 @@
+        elif isinstance(value, six.string_types):
+            self.String = value
+
+    def __eq__(self, other):
+        return other in [self.Function, self.String]
+
+
+class Equals(object):
+    """ Equals operator """
+    Left = None
+    Right = None
+
+    def __init__(self, equals):
+
+        if isinstance(equals, list):
+            if len(equals) == 2:
+                self.Left = EqualsValue(equals[0])
+                self.Right = EqualsValue(equals[1])
+
+    def test(self, scenarios):
+        """ Do an equals based on the provided scenario """
+        for scenario, value in scenarios.items():
+            if scenario == self.Left:
+                return value == self.Right
+            if scenario == self.Right:
+                return value == self.Left
+
+        return None
+
+
+class Condition(object):
+    """ Individual condition """
+    And = None
+    Or = None
+    Not = None
+    Equals = None
+    Influenced_Equals = None
+
+    def __init__(self, template, name):
+        self.And = []
+        self.Or = []
+        self.Not = []
+        self.Influenced_Equals = {}
+        value = template.get('Conditions', {}).get(name, {})
+        self.process_condition(template, value)
+
+    def test(self, scenarios):
+        """ Test a condition based on a scenario """
+        if self.And:
+            for a in self.And:
+                if not a.test(scenarios):
+                    return False
+            return True
+        if self.Or:
+            for o in self.Or:
+                if o.test(scenarios):
+                    return True
+            return False
+        if self.Not:
+            for n in self.Not:
+                return not n.test(scenarios)
+
+        return self.Equals.test(scenarios)
+
+    def process_influenced_equal(self, equal):
+        """ Get influenced equals from sub conditions """
+        if equal.Left.Function:
+            if not self.Influenced_Equals.get(equal.Left.Function):
+                self.Influenced_Equals[equal.Left.Function] = set()
+            if equal.Right.Function:
+                self.Influenced_Equals[equal.Left.Function].add(equal.Right.Function)
+            elif equal.Right.String:
+                self.Influenced_Equals[equal.Left.Function].add(equal.Right.String)
+        if equal.Right.Function:
+            if not self.Influenced_Equals.get(equal.Right.Function):
+                self.Influenced_Equals[equal.Right.Function] = set()
+            if equal.Left.Function:
+                self.Influenced_Equals[equal.Right.Function].add(equal.Left.Function)
+            elif equal.Left.String:
+                self.Influenced_Equals[equal.Right.Function].add(equal.Left.String)
+
+    def process_condition(self, template, value):
+        """ process condition """
+        if isinstance(value, dict):
+            if len(value) == 1:
+                for func_name, func_value in value.items():
+                    if func_name == 'Fn::And':
```

Comment:
To avoid typos, can we pull out the template tokens like `Fn::And` to CONST type values somewhere?

Reply:
Fair.  I can agree with that.